File: docs/content/game-theory/minimax/code/minimax.v1.full.cpp
/**
 * Minimax algorithm for two-player zero-sum games with perfect information.
 *
 * Finds the optimal move for the player to move, assuming the opponent also
 * plays optimally: one player maximizes the score, the other minimizes it.
 *
 * The algorithm is generic. A concrete game (see test/driver.cpp for the Coins
 * game) provides its own State type and two functions:
 *   - possible_moves_funct(state): the states reachable in one move; an empty
 *     vector means `state` is terminal (the game is over).
 *   - score_funct(state): the score of a terminal state, from the maximizing
 *     player's point of view (higher is better for them).
 *
 * A move is identified by its index in the possible_moves list, so best_move is
 * the index of the chosen child (-1 if `current_state` is already terminal).
 *
 * Returns: {best_value, best_move}.
 * Complexity: O(b^d) time, O(d) space (b = branching factor, d = tree depth).
 */
#include <vector>
#include <functional>
#include <limits>
#include <utility>
using namespace std;

template <typename State>
pair<double, int> minimax(
    const State& current_state,
    int maximizing_player,
    const function<vector<State>(const State&)>& possible_moves_funct,
    const function<double(const State&)>& score_funct) {

    // Base case: a terminal state (no moves left) is scored directly.
    vector<State> next_states = possible_moves_funct(current_state);
    if (next_states.empty())
        return {score_funct(current_state), -1};

    // The maximizing player wants the highest score, the minimizing one the lowest.
    double best_value = (maximizing_player == 1)
        ? -numeric_limits<double>::infinity()
        :  numeric_limits<double>::infinity();
    int best_move = -1;

    // Try every move; the opponent moves next, so flip who is maximizing.
    for (int index = 0; index < (int) next_states.size(); ++index) {
        double value = minimax(next_states[index], -maximizing_player,
                               possible_moves_funct, score_funct).first;
        if (maximizing_player == 1) {
            if (value > best_value) { best_value = value; best_move = index; }
        } else {
            if (value < best_value) { best_value = value; best_move = index; }
        }
    }

    return {best_value, best_move};
}

File: docs/content/game-theory/minimax/code/minimax.v1.full.py
"""Minimax algorithm for two-player zero-sum games with perfect information.

Finds the optimal move for the player to move, assuming the opponent also plays
optimally: one player maximizes the score while the other minimizes it.

The algorithm is generic — it knows nothing about a specific game. A concrete
game only provides two functions and an initial state (see `test/driver.py` for
the Coins game used in the example):

- `possible_moves_funct(state)`: the list of states reachable from `state` in one
  move. An empty list means `state` is terminal (the game is over).
- `score_funct(state)`: the score of a terminal `state`, from the maximizing
  player's point of view (higher is better for them).

A *move* is identified by its index in the list returned by
`possible_moves_funct`, so `best_move` is the index of the chosen child.

Args:
    current_state: the current game state (any value the game functions accept).
    maximizing_player (int): +1 if the player to move maximizes, -1 if minimizes.
    possible_moves_funct (callable): state -> list of next states.
    score_funct (callable): terminal state -> float score.

Returns:
    tuple[float, int]: the optimal score, and the index of the best move
    (None if `current_state` is already terminal).
"""

def minimax(current_state, maximizing_player: int, possible_moves_funct, score_funct) -> tuple[float, int]:

    # Base case: a terminal state (no moves left) is scored directly.
    next_states = possible_moves_funct(current_state)
    if not next_states:
        return score_funct(current_state), None

    # The maximizing player wants the highest score, the minimizing one the lowest.
    best_value = float('-inf') if maximizing_player == 1 else float('inf')
    best_move = None

    # Try every move; the opponent moves next, so flip who is maximizing.
    for index, next_state in enumerate(next_states):
        value, _ = minimax(next_state, -maximizing_player, possible_moves_funct, score_funct)

        if maximizing_player == 1:
            if value > best_value:
                best_value, best_move = value, index
        else:
            if value < best_value:
                best_value, best_move = value, index

    return best_value, best_move
